changed the hardcoded format of spreadsheet API

VinayCheripally opened the pull request:
**Description**

This PR fixes #703 

**Notes for Reviewers**


**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.
 

<!--
Thank you for contributing to Meshery! 

Contributing Conventions:

1. Include descriptive PR titles with [<component-name>] prepended.
2. Build and test your changes before submitting a PR. 
3. Sign your commits

By following the community's contribution conventions upfront, the review process will 
be accelerated and your PR merged more quickly.
-->

gourav-k-shaw commented:
LGTM.
@riyaa14 @lekaf974 you might have a feedback to offer

leecalcote commented:
Has this been tested with the current Meshery Integrations sheet?

lekaf974 reviewed (changes requested):
Added a comment with ideas to investigate

VinayCheripally commented:
@leecalcote I tested it and it is showing me that there zero models So  ,I'm trying to resolve that.

VinayCheripally commented:
@lekaf974 Yeah I think it makes sense to do what you are saying.I will work on them after I resolve this error with meshery Integration sheet.

VinayCheripally commented:
@leecalcote It seems the current URL format which is 
https://docs.google.com/spreadsheets/d/{spreadsheet-id}/gviz/tq?tqx=out:csv&gid={gid} is meant for visualization and is not meant for downloading the sheet as a csv file.
So I used the format https://docs.google.com/spreadsheets/d/{spreadsheet-id}/export?format=csv&gid={gid} which actually downloads the csv files.It works with both meshery Integrations sheet and also custom spreadsheets.
![Screenshot from 2025-04-03 13-57-33](https://github.com/user-attachments/assets/d9a95137-75bb-4b0e-884e-b31395e26d3b)
![Screenshot from 2025-04-03 13-59-19](https://github.com/user-attachments/assets/74141b33-3c98-4e81-ae87-7687bba66987)

VinayCheripally commented:
@lekaf974 I will add a test file to this.But I have a doubt using the default one with custom spreadsheets gives an error directly is that fine?

lekaf974 commented:
Can you share the error ?

riyaa14 commented:
> Do we want to create a PR each time we will have to change it (if needed), we modify an hardcoded value by another harcoded one. What about to have a default value (which should be the current one) and add it override by an argument in the mehod ? This will allow to not break the current behavior and let the caller to override it on runtime if needed

I strongly agree with this point. Since there is no official documentation for this new URL format, it is good to have a default value and override it if needed.

leecalcote commented:
Is there are link pointing to where Google Sheets has changed their URL format?

VinayCheripally commented:
> Is there are link pointing to where Google Sheets has changed their URL format?

I did not find an official article for that change in format but I found this blog.
https://youneedawiki.com/blog/posts/google-sheets-url-parameters.html

riyaa14 commented:
> Is there are link pointing to where Google Sheets has changed their URL format?

I couldn't find any documentation too regarding the change of URL format or even on what URL formats are valid. @VinayCheripally suggested some solutions which again aren't mentioned officially but rather from different sources on the net.

riyaa14 commented:
> But I have a doubt ,using the default one with custom spreadsheets gives an error directly is that fine?

@VinayCheripally actually this was the original problem. The URL format was working for Meshery Integrations sheet which I suppose is because it was published a long time back, but for recently published sheets, the URL wasn't working

and this is a problem as:
1. Users won't be able to use meshery registry commands for bulk model generation
2. E2E tests would need a new sheet too, we can't use the Meshery Integrations sheet here.

riyaa14 commented:
> This is the error we are getting right now when using custom spreadsheets.

@VinayCheripally what do you mean by custom spreadsheets here?

Also, if you see the url format you shared here, how is it again calling the original URL format : `/pub?output=csv&gid={sheet-id}`? Shouldn't this call the updated URL - `/export?format=csv&gid={sheet-id}` as per your changes in this PR.

lekaf974 commented:
also could we validate there is no impact on other command before merging

VinayCheripally commented:
> > This is the error we are getting right now when using custom spreadsheets.
> 
> @VinayCheripally what do you mean by custom spreadsheets here?
> 
> Also, if you see the url format you shared here, how is it again calling the original URL format : `/pub?output=csv&gid={sheet-id}`? Shouldn't this call the updated URL - `/export?format=csv&gid={sheet-id}` as per your changes in this PR.

I was showing this to @lekaf974 because he asked what kind of error was coming if we use the /pub?output=csv as default and override it by /export?format=csv if we use an argument. So,If we use this without argument in our tests for some test-spreadsheet it would give us an error and I asked @lekaf974 if that was fine or not.

VinayCheripally commented:
> > This is the error we are getting right now when using custom spreadsheets.
> 
> @VinayCheripally what do you mean by custom spreadsheets here?
> 
> Also, if you see the url format you shared here, how is it again calling the original URL format : `/pub?output=csv&gid={sheet-id}`? Shouldn't this call the updated URL - `/export?format=csv&gid={sheet-id}` as per your changes in this PR.

By custom spreadsheets I meant something which user created by his own. Not the Meshery Integrations sheet.

VinayCheripally commented:
https://developers.google.com/workspace/drive/api/reference/rest/v3/files/export
This is the closest I could find there is a somewhat similar that is being used here.

VinayCheripally commented:
@lekaf974 @riyaa14 I have a doubt currently in meshkit for publishing the models to some path we are using the url 
https://docs.google.com/spreadsheets/d/[spreadsheet-id]/pub?output=csv&gid=[Sheet-ID]
I believe this is used when we publish a spreadsheet to web which does not need credentials to access as we have already published it.So why are we using CRED variable in the mesheryctl command.
Even for the url format 
https://docs.google.com/spreadsheets/d/{spreadsheet-id}/export?format=csv&gid=0
This url works only if the spreadsheet can be accessed by anyone with the link.It does not use any authorization.
The only way which uses a .json file is the one with google sheets API which is not used directly to download the CSV file but to just read the file.So ,I'm confused why we are using the base64 format of .json file.

lekaf974 commented:
> @lekaf974 @riyaa14 I have a doubt currently in meshkit for publishing the models to some path we are using the url [https://docs.google.com/spreadsheets/d/[spreadsheet-id]/pub?output=csv&gid=[Sheet-ID]](https://docs.google.com/spreadsheets/d/%5Bspreadsheet-id%5D/pub?output=csv&gid=%5BSheet-ID%5D) I believe this is used when we publish a spreadsheet to web which does not need credentials to access as we have already published it.So why are we using CRED variable in the mesheryctl command. Even for the url format [https://docs.google.com/spreadsheets/d/{spreadsheet-id}/export?format=csv&gid=0](https://docs.google.com/spreadsheets/d/%7Bspreadsheet-id%7D/export?format=csv&gid=0) This url works only if the spreadsheet can be accessed by anyone with the link.It does not use any authorization. The only way which uses a .json file is the one with google sheets API which is not used directly to download the CSV file but to just read the file.So ,I'm confused why we are using the base64 format of .json file.

@VinayCheripally did not get involve in the command development but if you find some issues, create an issue with description then let's talk about it in slack and next dev meeting to get feedback.

I did not remind who worked on this implementation to give feedback, @leecalcote any idea ?

lekaf974 commented:
Pretty sure doing those e2e tests implementation will help detect issues that we were not aware due to small test coverage

riyaa14 commented:
> @lekaf974 @riyaa14 I have a doubt currently in meshkit for publishing the models to some path we are using the url [https://docs.google.com/spreadsheets/d/[spreadsheet-id]/pub?output=csv&gid=[Sheet-ID]](https://docs.google.com/spreadsheets/d/%5Bspreadsheet-id%5D/pub?output=csv&gid=%5BSheet-ID%5D) I believe this is used when we publish a spreadsheet to web which does not need credentials to access as we have already published it.So why are we using CRED variable in the mesheryctl command. Even for the url format [https://docs.google.com/spreadsheets/d/{spreadsheet-id}/export?format=csv&gid=0](https://docs.google.com/spreadsheets/d/%7Bspreadsheet-id%7D/export?format=csv&gid=0) This url works only if the spreadsheet can be accessed by anyone with the link.It does not use any authorization. The only way which uses a .json file is the one with google sheets API which is not used directly to download the CSV file but to just read the file.So ,I'm confused why we are using the base64 format of .json file.

@VinayCheripally spreadsheet-credentials (base64 encoded) are required for 
1. Getting the Sheet-IDs
2. In generate command, we update the components sheet and without credentials, we won't be able to edit the sheet

These are two areas where I found the use of CRED, might be more

VinayCheripally commented:
> https://developers.google.com/workspace/drive/api/reference/rest/v3/files/export This is the closest I could find about the URL format .there is a somewhat similar URL that is being explained here.

@riyaa14 Does this look similar?It has the same use case.can you look at it once?

VinayCheripally commented:
> @VinayCheripally
> 
> Some questions come to me
> 
>     1. We are making 3 times the same changes looks not a good practice IMO and not easy to maintain on long term
> 
>     2. Since it is in meshkit and it will be reused by external client (mesheryctl and server) to ensure it is not breaking them, I suggest to create a test
> 
>     3. Do we want to create a PR each time we will have to change it (if needed), we modify an hardcoded value by another harcoded one. What about to have a default value (which should be the current one) and add it override by an argument in the mehod ? This will allow to not break the current behavior and let the caller to override it on runtime if needed

I added a function in utils.go and used it at three places and by testing do you mean for only the function which I just changed or for all the functions in that model.go?

lekaf974 commented on registry/utils.go:
```diff
@@ -0,0 +1,7 @@
+import "fmt"
+
+func FormatSheetURL(baseURL string, sheetID int64) string {
+	return fmt.Sprintf("%s/export?format=csv&gid=%d", baseURL, sheetID)
```
I don;t really get it, FormatSheetURL return the same hardcoded value from a function so what we want to achieve ?

VinayCheripally replied:
As you said we were making the same changes three times , I thought using a function would make it easier to maintain?

riyaa14 replied:
@VinayCheripally we can define a variable in `registry/helper.go` for this url. and this variable can be used at all the places. A function doesn't seem like the right choice here.

lekaf974 replied:
+1 about last point.

You identified the issue, I was saying to keep functionallity as is and just allow url overriding where the current value doesn't work with what we already have. 

PS: Perhaps we can make it a little bit cleaner but it can be done in another issue.

lekaf974 commented on registry/helpers.go:
```diff
@@ -20,6 +20,8 @@ const (
 	artifactHub            = "artifacthub"
 	gitHub                 = "github"
 	shouldRegisterColIndex = -1
+	defaultURL             = "/pub?output=csv&gid="
+	overrideURL            = "/export?format=csv&gid="
```
```suggestion
	defaultURLPathAndQueryParams             = "/pub?output=csv&gid="             = "/pub?output=csv&gid="
	overridedtURLPathAndQueryParams       = "/pub?output=csv&gid="            = "/export?format=csv&gid="
```

lekaf974 reviewed (changes requested):
Renaming proposal also how it will modify the behavior on the runtime when mesheryctl will trigger one of those method ?

VinayCheripally commented:
> Renaming proposal also how it will modify the behavior on the runtime when mesheryctl will trigger one of those method ?

So whenever the defaultURI returns an error from the downloadFile function it uses the overrideURL and calls the same function again and checks for errors and just returns if it works as intended or else returns the error.

VinayCheripally commented:
> > > Renaming proposal also how it will modify the behavior on the runtime when mesheryctl will trigger one of those method ?
> > 
> > 
> > So whenever the defaultURI returns an error from the downloadFile function it uses the overrideURL and calls the same function again and checks for errors and just returns if it works as intended or else returns the error.
> 
> Ok, did not realized it on first read, in this case here it is important to add a comment this behavior explaining why we do this, please.

got it ,I'll do that.

VinayCheripally commented:
@lekaf974 I added the comments.

lekaf974 commented on registry/relationship.go:
```diff
@@ -47,14 +47,19 @@ type RelationshipCSV struct {
+		err := utils.DownloadFile(csvPath, newSheetURL)
 		if err != nil {
-			return nil, utils.ErrReadingRemoteFile(err)
+			// If downloading with the default URL Path fails, retry using the overridden URL Path.
```
The comment says what the code is doing so it is not use full, please add context on why we have to do this patch

lekaf974 replied:
I mean we did this fix because some command were broken in mesheryctl, etc...

VinayCheripally replied:
There are two commands which are broken because of this (registry publish and registry generate). I thought rather than adding the commands adding the core reason that why they are broken would be better, but yeah, I will add the commands just in case.

VinayCheripally replied:
Added the commands in comments!

lekaf974 commented on registry/component.go:
```diff
@@ -209,14 +209,21 @@ type ComponentCSVHelper struct {
+		err := utils.DownloadFile(csvPath, newSheetURL)
 		if err != nil {
-			return nil, utils.ErrReadingRemoteFile(err)
+			// The URL format /pub?output=csv (defaultURLPathAndQueryParams) fails for spreadsheets not published to the web.
+			// mesheryctl registry publish and mesheryctl registry generate might fail If spreadsheets are not published.
+			// Instead, we use the /export?format=csv format (overridedURLPathAndQueryParams) which works for sheets that are shared with everyone.
```
@VinayCheripally I had in mind this logic was embedded in a reusable function ?

VinayCheripally replied:
@lekaf974 It's not embedded in a reusable function .Should I create a function in helpers.go and use it in all these three files?

VinayCheripally replied:
@riyaa14 A doubt. Using the format /pub is not a deprecation right because it's meant for publishing it's just a bug right.

VinayCheripally replied:
ok... didn't know that. I will update the comments.

lekaf974 commented on registry/helpers.go:
```diff
@@ -52,3 +56,26 @@ func SetupLogger(name string, debugLevel bool, output io.Writer) logger.Handler
+		if err !=nil{
+			return "",utils.ErrReadingRemoteFile(err)
+		}
+	}
```
+1

VinayCheripally commented on registry/helpers.go:
```diff
@@ -52,3 +56,26 @@ func SetupLogger(name string, debugLevel bool, output io.Writer) logger.Handler
+// Downloads CSV file using spreadsheet URL
+func DownloadCSV(sheetURL ,csvPath string, spreadsheetID int64)  (string ,error){
+	newSheetURL := sheetURL + defaultURLPathAndQueryParams + strconv.FormatInt(spreadsheetID, 10)
+	Log.Info("Downloading CSV from: ", newSheetURL)
```
Yup that makes sense.

VinayCheripally commented on registry/relationship.go:
```diff
@@ -47,14 +47,18 @@ type RelationshipCSV struct {
 	if localCsvPath == "" {
-		sheetURL = sheetURL + "/pub?output=csv" + "&gid=" + strconv.FormatInt(spreadsheetID, 10)
-		Log.Info("Downloading CSV from: ", sheetURL)
+		newSheetURL := sheetURL + defaultURLPathAndQueryParams + strconv.FormatInt(spreadsheetID, 10)
+		Log.Info("Downloading CSV from: ", newSheetURL)
```
Yeah removing it.

lekaf974 commented on registry/helpers.go:
```diff
@@ -52,3 +56,26 @@ func SetupLogger(name string, debugLevel bool, output io.Writer) logger.Handler
 }
+
+// Downloads CSV file using spreadsheet URL
+func DownloadCSV(sheetURL ,csvPath string, spreadsheetID int64)  (string ,error){
```
Just seen this why a Method called DownloadCSV is returning newSheetUrl ? the name is misleading

lekaf974 replied:
Also wondering if we are not downloading twice, one time to retrieve the sheet url in the url and another time when calling in each files

VinayCheripally replied:
@riyaa14 @lekaf974 newSheetURL is being assigned to the variable sheetURL which is then used as spreadsheetURL while returning CSV helpers.
https://github.com/meshery/meshkit/blob/0bf035aa3b2328b0b8fbf217458a0dbf9f74a144/registry/model.go#L236

VinayCheripally replied:
@lekaf974 Are you talking about how we are calling the functions from meshery repo ?

lekaf974 replied:
So like @riyaa14 mentioned remove the string returned since it is not used anymore

VinayCheripally replied:
> @riyaa14 @lekaf974 newSheetURL is being assigned to the variable sheetURL which is then used as spreadsheetURL while returning CSV helpers.
> 
> https://github.com/meshery/meshkit/blob/0bf035aa3b2328b0b8fbf217458a0dbf9f74a144/registry/model.go#L236

@lekaf974 Here I already mentioned that it is being used as spreadsheetURL in initialized ModelCSVHelper.

lekaf974 commented:
@riyaa14 last check ok to merge then
